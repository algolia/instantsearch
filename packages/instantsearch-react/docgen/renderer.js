import marked from 'marked';
import {runMode} from 'codemirror/addon/runmode/runmode.node';
import 'codemirror/mode/jsx/jsx';
import escape from 'escape-html';

const highlight = source => {
  let output = '';
  // Even though highlight accepts a lang parameter, force to JSX for now.
  runMode(source, 'jsx', (text, style) => {
    text = escape(text);
    if (!style) {
      output += text;
      return;
    }
    const className = style.split(' ').map(s => `cm-${s}`).join(' ');
    output += `<span class="${className}">${text}</span>`;
  });
  return output;
};

const blockTheme = 'cm-s-default';
const spanTheme = 'cm-s-default';
const renderer = new marked.Renderer();
renderer.code = (code, lang) => {
  return `<pre class="CodeMirror ${blockTheme}"><code>${highlight(code, lang)}</code></pre>`;
};
renderer.codespan = code => {
  return `<code class="CodeMirror ${spanTheme}">${code}</code>`;
};

export default renderer;
