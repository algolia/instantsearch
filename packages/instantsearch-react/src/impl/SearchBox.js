import React, {Component, PropTypes} from 'react';
import themeable from 'react-themeable';

import {getTranslation} from './utils';

const defaultTranslations = {
  submit: null,
  reset: null,
  submitTitle: 'Submit your search query.',
  resetTitle: 'Clear the search query.',
  placeholder: 'Search your website',
};

const defaultTheme = {
  root: 'SearchBox',
  wrapper: 'SearchBox__wrapper',
  input: 'SearchBox__input',
  submit: 'SearchBox__submit',
  reset: 'SearchBox__reset',
};

class SearchBox extends Component {
  static propTypes = {
    query: PropTypes.string,
    refine: PropTypes.func.isRequired,
    theme: PropTypes.object,
    translations: PropTypes.object,
    // @TODO: implement
    // poweredBy: PropTypes.bool,
    focusShortcuts: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ),
    autoFocus: PropTypes.bool,
    searchAsYouType: PropTypes.bool,
  };

  static defaultProps = {
    query: '',
    theme: defaultTheme,
    translations: defaultTranslations,
    poweredBy: false,
    focusShortcuts: ['s', '/'],
    autoFocus: false,
    searchAsYouType: true,
  };

  constructor(props) {
    super();

    this.state = {
      query: props.searchAsYouType ? null : props.query,
    };
  }

  componentDidMount() {
    document.addEventListener('keydown', this.onKeyDown);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown);
  }

  componentWillReceiveProps(nextProps) {
    // Reset query when the searchParameters query has changed.
    // This is kind of an anti-pattern (props in state), but it works here
    // since we know for sure that searchParameters having changed means a
    // new search has been triggered.
    if (
      !nextProps.searchAsYouType &&
      nextProps.query !== this.props.query
    ) {
      this.setState({
        query: nextProps.query,
      });
    }
  }

  getQuery = () =>
    this.props.searchAsYouType ?
      this.props.query :
      this.state.query;

  setQuery = val => {
    const {refine, searchAsYouType} = this.props;
    if (searchAsYouType) {
      refine(val);
    } else {
      this.setState({
        query: val,
      });
    }
  }

  onInputMount = input => {
    this.input = input;
  };

  // From https://github.com/algolia/autocomplete.js/pull/86
  onKeyDown = e => {
    if (!this.props.focusShortcuts) {
      return;
    }

    const shortcuts = this.props.focusShortcuts.map(key =>
      typeof key === 'string' ? key.toUpperCase().charCodeAt(0) : key
    );

    const elt = e.target || e.srcElement;
    const tagName = elt.tagName;
    if (
      elt.isContentEditable ||
      tagName === 'INPUT' ||
      tagName === 'SELECT' ||
      tagName === 'TEXTAREA'
    ) {
      // already in an input
      return;
    }

    const which = e.which || e.keyCode;
    if (shortcuts.indexOf(which) === -1) {
      // not the right shortcut
      return;
    }

    this.input.focus();
    e.stopPropagation();
    e.preventDefault();
  };

  onSubmit = e => {
    e.preventDefault();
    e.stopPropagation();

    const {refine, searchAsYouType} = this.props;
    if (!searchAsYouType) {
      refine(this.getQuery());
    }
    return false;
  };

  onChange = e => {
    this.setQuery(e.target.value);
  };

  onReset = () => {
    this.setQuery('');
    this.input.focus();
  };

  render() {
    const {
      theme,
      translations,
      autoFocus,
    } = this.props;
    const query = this.getQuery();
    const th = themeable(theme);

    return (
      <form
        noValidate
        onSubmit={this.onSubmit}
        onReset={this.onReset}
        {...th('root', 'root')}
      >
        <div
          role="search"
          {...th('wrapper', 'wrapper')}
        >
          <input
            ref={this.onInputMount}
            type="search"
            placeholder={getTranslation(
              'placeholder',
              defaultTranslations,
              translations
            )}
            autoFocus={autoFocus}
            autoComplete={false}
            required
            value={query}
            onChange={this.onChange}
            {...th('input', 'input')}
          />
          <button
            type="submit"
            title={getTranslation(
              'submitTitle',
              defaultTranslations,
              translations
            )}
            {...th('submit', 'submit')}
          >
            {getTranslation('submit', defaultTranslations, translations)}
          </button>
          <button
            type="reset"
            title={getTranslation(
              'resetTitle',
              defaultTranslations,
              translations
            )}
            {...th('reset', 'reset')}
          >
            {getTranslation('reset', defaultTranslations, translations)}
          </button>
        </div>
      </form>
    );
  }
}

export default SearchBox;
