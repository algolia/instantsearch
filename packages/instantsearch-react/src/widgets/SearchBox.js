import createSearchBox from '../hoc/createSearchBox';
import SearchBoxImpl from '../impl/SearchBox';
export default createSearchBox(SearchBoxImpl);
