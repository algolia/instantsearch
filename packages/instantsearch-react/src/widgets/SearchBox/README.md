---
title: SearchBox
layout: api.ejs
nav_groups:
  - widgets
---

# SearchBox

The `SearchBox` widget is where users type their search queries.

## Props

<!-- props default ./index -->

### Theme

`root`, `wrapper`, `input`, `submit`, `reset`

### Translations

`submit`, `reset`, `submitTitle`, `resetTitle`, `placeholder`

## Implementing your own SearchBox

See [Making your own widgets](../Customization.md) for more information on how to use the `SearchBox.connect` HOC.

```
import {SearchBox.connect} from 'instantsearch-react';

function MySearchBox(props) {
  return (
    <input
      value={props.query}
      onChange={e => props.refine(e.target.value)}
    />
  );
}

// `SearchBox.connect` accepts the same `id` prop as `SearchBox`.
export default SearchBox.connect(MySearchBox);
```
