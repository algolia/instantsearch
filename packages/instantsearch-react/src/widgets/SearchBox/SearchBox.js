import React, {Component, PropTypes} from 'react';

import themeable from '../../core/themeable';
import translatable from '../../core/translatable';

class SearchBox extends Component {
  static propTypes = {
    query: PropTypes.string,
    refine: PropTypes.func.isRequired,
    applyTheme: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,
    // @TODO: implement
    // poweredBy: PropTypes.bool,

    /**
     * List of keyboard shortcuts that focus the search box.
     * Accepts key names and key codes.
     * @public
     */
    focusShortcuts: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ),

    /**
     * Should the search box be focused on render?
     * @public
     */
    autoFocus: PropTypes.bool,

    /**
     * Should we search on every change to the query?
     * If you disable this option, new searches will only be triggered by
     * clicking the search button or by pressing the enter key while the search
     * box is focused.
     * @public
     */
    searchAsYouType: PropTypes.bool,

    // For testing purposes
    __inputRef: PropTypes.func,
  };

  static defaultProps = {
    query: '',
    // poweredBy: false,
    focusShortcuts: ['s', '/'],
    autoFocus: false,
    searchAsYouType: true,
  };

  constructor(props) {
    super();

    this.state = {
      query: props.searchAsYouType ? null : props.query,
    };
  }

  componentDidMount() {
    document.addEventListener('keydown', this.onKeyDown);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown);
  }

  componentWillReceiveProps(nextProps) {
    // Reset query when the searchParameters query has changed.
    // This is kind of an anti-pattern (props in state), but it works here
    // since we know for sure that searchParameters having changed means a
    // new search has been triggered.
    if (
      !nextProps.searchAsYouType &&
      nextProps.query !== this.props.query
    ) {
      this.setState({
        query: nextProps.query,
      });
    }
  }

  getQuery = () =>
    this.props.searchAsYouType ?
      this.props.query :
      this.state.query;

  setQuery = val => {
    const {refine, searchAsYouType} = this.props;
    if (searchAsYouType) {
      refine(val);
    } else {
      this.setState({
        query: val,
      });
    }
  }

  onInputMount = input => {
    this.input = input;
    if (this.props.__inputRef) {
      this.props.__inputRef(input);
    }
  };

  // From https://github.com/algolia/autocomplete.js/pull/86
  onKeyDown = e => {
    if (!this.props.focusShortcuts) {
      return;
    }

    const shortcuts = this.props.focusShortcuts.map(key =>
      typeof key === 'string' ? key.toUpperCase().charCodeAt(0) : key
    );

    const elt = e.target || e.srcElement;
    const tagName = elt.tagName;
    if (
      elt.isContentEditable ||
      tagName === 'INPUT' ||
      tagName === 'SELECT' ||
      tagName === 'TEXTAREA'
    ) {
      // already in an input
      return;
    }

    const which = e.which || e.keyCode;
    if (shortcuts.indexOf(which) === -1) {
      // not the right shortcut
      return;
    }

    this.input.focus();
    e.stopPropagation();
    e.preventDefault();
  };

  onSubmit = e => {
    e.preventDefault();
    e.stopPropagation();

    const {refine, searchAsYouType} = this.props;
    if (!searchAsYouType) {
      refine(this.getQuery());
    }
    return false;
  };

  onChange = e => {
    this.setQuery(e.target.value);
  };

  onReset = () => {
    this.setQuery('');
    this.input.focus();
  };

  render() {
    const {
      applyTheme,
      translate,
      autoFocus,
    } = this.props;
    const query = this.getQuery();

    return (
      <form
        noValidate
        onSubmit={this.onSubmit}
        onReset={this.onReset}
        {...applyTheme('root', 'root')}
      >
        <div
          role="search"
          {...applyTheme('wrapper', 'wrapper')}
        >
          <input
            ref={this.onInputMount}
            type="search"
            placeholder={translate('placeholder')}
            autoFocus={autoFocus}
            autoComplete={false}
            required
            value={query}
            onChange={this.onChange}
            {...applyTheme('input', 'input')}
          />
          <button
            type="submit"
            title={translate('submitTitle')}
            {...applyTheme('submit', 'submit')}
          >
            {translate('submit')}
          </button>
          <button
            type="reset"
            title={translate('resetTitle')}
            {...applyTheme('reset', 'reset')}
          >
            {translate('reset')}
          </button>
        </div>
      </form>
    );
  }
}

export default themeable({
  root: 'SearchBox',
  wrapper: 'SearchBox__wrapper',
  input: 'SearchBox__input',
  submit: 'SearchBox__submit',
  reset: 'SearchBox__reset',
})(
  translatable({
    submit: null,
    reset: null,
    submitTitle: 'Submit your search query.',
    resetTitle: 'Clear the search query.',
    placeholder: 'Search your website',
  })(SearchBox)
);
