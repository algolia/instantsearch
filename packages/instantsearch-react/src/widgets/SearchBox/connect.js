import {PropTypes} from 'react';

import createConnector from '../../core/createConnector';

function getQuery(props, state) {
  if (typeof state[props.id] !== 'undefined') {
    return state[props.id];
  }
  return '';
}

export default createConnector({
  displayName: 'AlgoliaSearchBox',

  propTypes: {
    /**
     * URL state serialization key.
     * The state of this widget takes the form of a `string`.
     * @public
     */
    id: PropTypes.string,
  },

  defaultProps: {
    id: 'q',
  },

  getProps(props, state) {
    return {
      query: getQuery(props, state),
    };
  },

  refine(props, state, nextQuery) {
    return {
      ...state,
      [props.id]: nextQuery,
    };
  },

  getSearchParameters(searchParameters, props, state) {
    return searchParameters.setQuery(getQuery(props, state));
  },

  getMetadata(props, state) {
    const query = getQuery(props, state);
    return {
      id: props.id,
      filters: !query ? [] : [{
        key: props.id,
        label: query,
        hide: true,
        clear: nextState => ({...nextState, [props.id]: ''}),
      }],
    };
  },
});
