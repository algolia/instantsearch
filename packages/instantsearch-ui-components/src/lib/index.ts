export * from './cx';
export * from './stickToBottom';
export * from './utils';
