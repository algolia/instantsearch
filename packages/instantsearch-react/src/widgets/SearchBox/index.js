import connect from './connect';
import SearchBox from './SearchBox';

const Connected = connect(SearchBox);
Connected.connect = connect;
export default Connected;
