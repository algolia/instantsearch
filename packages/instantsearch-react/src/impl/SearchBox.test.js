/* eslint-env jest, jasmine */

import React from 'react';
import {findDOMNode} from 'react-dom';
import {mount} from 'enzyme';
import renderer from 'react/lib/ReactTestRenderer';

import SearchBox from './SearchBox';
jest.unmock('./SearchBox');
jest.unmock('./utils');

let tree;

// In the current version of React, adding refs to elements will make
// ReactTestRenderer throw.
// Fixed in https://github.com/facebook/react/commit/caec8d5ce719997c15a4d586ceb771fcfbb5ccf7#diff-627ac9392c784e6a37db96c7c8efb219R66
// In the meantime, we can monkeypatch ReactTestComponent by devious means to
// add this patch... (pls don't fire me)
import {createInternalComponent} from 'react/lib/ReactHostComponent';
const dummy = createInternalComponent(<div />);
dummy.constructor.prototype.getPublicInstance = () => null;

describe('SearchBox', () => {
  it('applies its default props', () => {
    tree = renderer.create(
      <SearchBox />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('transfers the autoFocus prop to the underlying input element', () => {
    tree = renderer.create(
      <SearchBox autoFocus />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('treats its query prop as its input value', () => {
    // @TODO: Once ReactTestRenderer supports updating components, we'll be
    // able to use snapshots for this.
    const wrapper = mount(
      <SearchBox query="QUERY1" />
    );
    expect(wrapper.find('input').props().value).toBe('QUERY1');
    wrapper.setProps({
      query: 'QUERY2',
    });
    expect(wrapper.find('input').props().value).toBe('QUERY2');
    wrapper.unmount();
  });

  it('lets you customize its theme', () => {
    tree = renderer.create(
      <SearchBox
        theme={{
          root: 'ROOT',
          wrapper: 'WRAPPER',
          input: 'INPUT',
          submit: 'SUBMIT',
          reset: 'RESET',
        }}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('lets you customize its translations', () => {
    tree = renderer.create(
      <SearchBox
        translations={{
          submit: 'SUBMIT',
          reset: 'RESET',
          submitTitle: 'SUBMIT_TITLE',
          resetTitle: 'RESET_TITLE',
          placeholder: 'PLACEHOLDER',
        }}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('treats query as a default value when searchAsYouType=false', () => {
    const wrapper = mount(
      <SearchBox query="QUERY1" searchAsYouType={false} />
    );
    expect(wrapper.find('input').props().value).toBe('QUERY1');
    wrapper.find('input').simulate('change', {target: {value: 'QUERY2'}});
    expect(wrapper.find('input').props().value).toBe('QUERY2');
    wrapper.unmount();
  });

  it('overrides its value on query changes when searchAsYouType=false', () => {
    const wrapper = mount(
      <SearchBox query="QUERY1" searchAsYouType={false} />
    );
    expect(wrapper.find('input').props().value).toBe('QUERY1');
    wrapper.setProps({
      query: 'QUERY2',
    });
    expect(wrapper.find('input').props().value).toBe('QUERY2');
    wrapper.unmount();
  });

  it('refines its value on change when searchAsYouType=true', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <SearchBox
        searchAsYouType
        refine={refine}
      />
    );
    wrapper.find('input').simulate('change', {target: {value: 'hello'}});
    expect(refine.mock.calls.length).toBe(1);
    expect(refine.mock.calls[0][0]).toBe('hello');
    wrapper.unmount();
  });

  it('only refines its query on submit when searchAsYouType=false', () => {
    const refine = jest.fn();
    const wrapper = mount(
      <SearchBox
        searchAsYouType={false}
        refine={refine}
      />
    );
    wrapper.find('input').simulate('change', {target: {value: 'hello'}});
    expect(refine.mock.calls.length).toBe(0);
    wrapper.find('form').simulate('submit');
    expect(refine.mock.calls.length).toBe(1);
    expect(refine.mock.calls[0][0]).toBe('hello');
    wrapper.unmount();
  });

  it('focuses the input when one of the keys in focusShortcuts is pressed', () => {
    const wrapper = mount(
      <SearchBox
        focusShortcuts={['s', 84]}
      />
    );
    const input = findDOMNode(wrapper.instance()).querySelector('input');
    input.focus = jest.fn();
    // Note that all our `render.create` made components aren't unmounted
    // (the API hasn't made it in yet) so they will also receive the keydown
    // event. Events for everyone!
    const event1 = new KeyboardEvent('keydown', {keyCode: 82});
    document.dispatchEvent(event1);
    expect(input.focus.mock.calls.length).toBe(0);
    const event2 = new KeyboardEvent('keydown', {keyCode: 83});
    document.dispatchEvent(event2);
    expect(input.focus.mock.calls.length).toBe(1);
    const event3 = new KeyboardEvent('keydown', {keyCode: 84});
    document.dispatchEvent(event3);
    expect(input.focus.mock.calls.length).toBe(2);
  });
});
