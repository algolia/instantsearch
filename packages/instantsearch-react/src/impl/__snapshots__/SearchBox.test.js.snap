exports[`SearchBox applies its default props 0`] = `
<form
  className="SearchBox"
  noValidate={true}
  onReset={[Function anonymous]}
  onSubmit={[Function anonymous]}>
  <div
    className="SearchBox__wrapper"
    role="search">
    <input
      autoComplete={false}
      autoFocus={false}
      className="SearchBox__input"
      onChange={[Function anonymous]}
      placeholder="Search your website"
      required={true}
      type="search"
      value="" />
    <button
      className="SearchBox__submit"
      title="Submit your search query."
      type="submit" />
    <button
      className="SearchBox__reset"
      title="Clear the search query."
      type="reset" />
  </div>
</form>
`;

exports[`SearchBox lets you customize its theme 0`] = `
<form
  className="ROOT"
  noValidate={true}
  onReset={[Function anonymous]}
  onSubmit={[Function anonymous]}>
  <div
    className="WRAPPER"
    role="search">
    <input
      autoComplete={false}
      autoFocus={false}
      className="INPUT"
      onChange={[Function anonymous]}
      placeholder="Search your website"
      required={true}
      type="search"
      value="" />
    <button
      className="SUBMIT"
      title="Submit your search query."
      type="submit" />
    <button
      className="RESET"
      title="Clear the search query."
      type="reset" />
  </div>
</form>
`;

exports[`SearchBox lets you customize its translations 0`] = `
<form
  className="SearchBox"
  noValidate={true}
  onReset={[Function anonymous]}
  onSubmit={[Function anonymous]}>
  <div
    className="SearchBox__wrapper"
    role="search">
    <input
      autoComplete={false}
      autoFocus={false}
      className="SearchBox__input"
      onChange={[Function anonymous]}
      placeholder="PLACEHOLDER"
      required={true}
      type="search"
      value="" />
    <button
      className="SearchBox__submit"
      title="SUBMIT_TITLE"
      type="submit">
      SUBMIT
    </button>
    <button
      className="SearchBox__reset"
      title="RESET_TITLE"
      type="reset">
      RESET
    </button>
  </div>
</form>
`;

exports[`SearchBox transfers the autoFocus prop to the underlying input element 0`] = `
<form
  className="SearchBox"
  noValidate={true}
  onReset={[Function anonymous]}
  onSubmit={[Function anonymous]}>
  <div
    className="SearchBox__wrapper"
    role="search">
    <input
      autoComplete={false}
      autoFocus={true}
      className="SearchBox__input"
      onChange={[Function anonymous]}
      placeholder="Search your website"
      required={true}
      type="search"
      value="" />
    <button
      className="SearchBox__submit"
      title="Submit your search query."
      type="submit" />
    <button
      className="SearchBox__reset"
      title="Clear the search query."
      type="reset" />
  </div>
</form>
`;
